### PR TITLE
Doc 9058 and 9181: document support for auto failover of index service; updates for graceful and concurrent failover

### PR DESCRIPTION
**Pull request by RayOffiah**

### CODE REVIEW

**Comment by menih** [comment]
* We need to cover co-located data/index services. data service state always takes precedence over indexer.
* Indexer failover safety will be best effort based on the health state reported at the time we decide to failover.

**Comment by menih** [comment]
@vzasade please review and add your comments

**Comment by RayOffiah** [comment]
> * We need to cover co-located data/index services. data service state always takes precedence over indexer.
> * Indexer failover safety will be best effort based on the health state reported at the time we decide to failover.

I think this might have been covered, but I'll add a note to make it clearer. How about this:

image

**Comment by RayOffiah** [comment]
Would be great if @jeelanp2003 and/or @menih  could have another quick look at the changes.

**Comment by RayOffiah** [comment]
Awaiting further reviews.

**Review by cherkauer-couchbase — COMMENTED** [review]
AFAIK this is correct but really ns_server needs to review because that is the component making the actual Autofailover decisions.

@vzasade would be the correct person for that.

**Review by cherkauer-couchbase — COMMENTED** [review]
"Disk read/write failure. Attempts to read from or write to disk on a particular node have resulted in a significant rate of failure, for longer than a specified time-period. The node is removed by auto-failover, even though the node continues to be contactable."

This part is not yet implemented for GSI.

File: modules/learn/pages/clusters-and-availability/automatic-failover.adoc [hunk 1]
```diff
@@ -83,9 +83,9 @@ The node-minimum for each service is provided in the following table:
 
 | Query
 | 2
-
+// TODO: Check the number for the index.
 | Index
-| Not supported
+| 3
```
Comment (vzasade) [inline modules/learn/pages/clusters-and-availability/automatic-failover.adoc]: 2

File: modules/learn/pages/clusters-and-availability/automatic-failover.adoc [hunk 2]
```diff
@@ -108,10 +108,10 @@ This is illustrated by the following examples:
 | Node | Services Hosted
 
 | #1
-| Data
+| Data & Index
```
Comment (vzasade) [inline modules/learn/pages/clusters-and-availability/automatic-failover.adoc]: I don't think including Index in this example is needed

File: modules/learn/pages/clusters-and-availability/automatic-failover.adoc [hunk 3]
```diff
@@ -171,7 +171,7 @@ In this case, even though the Query and Search Services were both running on onl
 If node #1, #2, or #3 becomes unresponsive, auto-failover can be triggered.
 In each case, this is due to _Data Service Preference_, which applies auto-failover based on the policy for the Data Service, irrespective of other services on the unresponsive node.
 Note that in the case of node #2, this allows an Index Service node to be automatically failed over.
-However, if node #4 becomes unresponsive, auto-failover is not triggered; since auto-failover is not supported for the Index Service, unless Data Service Preference applies.
+If node #4 becomes unresponsive, auto-failover is triggered as there are four nodes (The threshold for the index service is 3.)
```
Comment (vzasade) [inline modules/learn/pages/clusters-and-availability/automatic-failover.adoc]: threshold for index service is 2, as for all other services

**Comment by vzasade** [comment]
>> Group failure. An administrator-defined group
We removed that feature in Neo. Instead the autofailover can fail over multiple nodes at the same time regardless of them belonging to configured group

**Comment by vzasade** [comment]
>> Only on the occurrence of one event at a time

Not true anymore. If multiple nodes are down at the same time, we will fail over multiple nodes, till the maximum of allowed failed over nodes is reached.

**Comment by vzasade** [comment]
>> The highest permitted maximum is 3

We are removing this limit

File: modules/learn/pages/clusters-and-availability/automatic-failover.adoc (hunk 2 above)
Comment (RayOffiah) [reply modules/learn/pages/clusters-and-availability/automatic-failover.adoc]: Removed & Index

**Review by cherkauer-couchbase — COMMENTED** [review]
"Note that in the case of node #2, this allows an Index Service node to be automatically failed over."

This verbiage that implies this is the only way for Index to be automatically failed over is leftover from when that was the case. Maybe should be tweaked or just removed now that Index can fail over on its own.

File: modules/learn/pages/clusters-and-availability/automatic-failover.adoc [hunk 4]
```diff
@@ -1,7 +1,7 @@
 = Automatic Failover
 :description: A node or group can be failed over automatically when it either becomes unresponsive or experiences continuous disk-access problems.
```
Comment (shivaniguptacb) [inline modules/learn/pages/clusters-and-availability/automatic-failover.adoc]: Replace with: One or more nodes, up to a maximum of less than half the cluster (i.e. (nodes +1)/2 -1 ), can be failed over automatically when the nodes become unresponsive or experience continuous disk problems.

**Review by shivaniguptacb — COMMENTED** [review]
Overall the changes look good. Just have one comment for the overall description of Automatic Failover.

**Review by shivaniguptacb — CHANGES_REQUESTED** [review]
See the change for the headline of Automatic Failoverr.

**Review by tonyjhillman — APPROVED** [review]

File: modules/learn/pages/clusters-and-availability/automatic-failover.adoc [hunk 5]
```diff
@@ -19,26 +20,16 @@ For information on managing auto-failover, see the information provided for Couc
...
 
 [#auto-failover-constraints]
 == Auto-Failover Constraints
 
 Auto-failover is triggered:
 
-* Only on the occurrence of one event at a time.
-If multiple events are concurrent, auto-failover is not triggered.
+* If the majority of nodes (stem:[(nodes + 1)/2]) in the cluster can form a quorum to initiate failover following the failure of one the nodes.
 * Sequentially, only up to an administrator-specified maximum number of events.
```
Comment (shivaniguptacb) [inline modules/learn/pages/clusters-and-availability/automatic-failover.adoc:32]: Change to 'Sequentially, only up to an administrator-specified maximum number of nodes.'

File: modules/learn/pages/clusters-and-availability/automatic-failover.adoc [hunk 6]
```diff
@@ -19,26 +20,16 @@ For information on managing auto-failover, see the information provided for Couc
...
-The highest permitted maximum is 3.
```
Comment (shivaniguptacb) [inline modules/learn/pages/clusters-and-availability/automatic-failover.adoc:41]: This line should be removed. We don't have this limit anymore.

File: modules/learn/pages/clusters-and-availability/automatic-failover.adoc [hunk 7]
```diff
@@ -19,26 +20,16 @@ For information on managing auto-failover, see the information provided for Couc
...
+The autofailover can fail over multiple nodes at the same time.
 
 [#auto-failover-constraints]
 == Auto-Failover Constraints
 
 Auto-failover is triggered:
 
-* Only on the occurrence of one event at a time.
-If multiple events are concurrent, auto-failover is not triggered.
+* If the majority of nodes (stem:[(nodes + 1)/2]) in the cluster can form a quorum to initiate failover following the failure of one the nodes.
```
Comment (shivaniguptacb) [inline modules/learn/pages/clusters-and-availability/automatic-failover.adoc:31]: Modify to:
If the majority of nodes (stem:[(nodes + 1)/2]) in the cluster can form a quorum to initiate failover following the failure of one or more nodes. Multiple concurrently failing nodes will be auto-failed over as long as the administrator specified limit is not exceeded.

File: modules/learn/pages/clusters-and-availability/automatic-failover.adoc (hunk 4 above)
Comment (shivaniguptacb) [reply modules/learn/pages/clusters-and-availability/automatic-failover.adoc]: One or more nodes can be failed over automatically when the nodes become unresponsive or experience continuous disk problems. This is done as long as data safety is maintained i.e. no data will be lost.

**Comment by RayOffiah** [comment]
This one's a bit of a mess. Closing without merging

**Pull request closed without merging**